src/economy/economy.cpp

```diff
@@ -13,33 +13,591 @@
+		auto sid = state.world.state_ownership_get_state(soid);
+		auto market = state.world.state_instance_get_market_from_local_market(sid);
+
+		auto local = state.world.market_get_demand(market, c);
+		total += local * state.world.market_get_demand_satisfaction(market, c);
+	});
+
+	return total;
+}
+float consumption(
+	sys::state& state,
+	dcon::commodity_id c
+) {
+	auto total = 0.f;
+	state.world.for_each_market([&](auto m) {
+		auto local = state.world.market_get_demand(m, c);
+		total += local * state.world.market_get_demand_satisfaction(m, c);
+	});
+	return total;
+}
+
+float market_pool(
+	sys::state& state,
+	dcon::market_id s,
+	dcon::commodity_id c
+) {
+	return state.world.market_get_stockpile(s, c);
+}
+float market_pool(
+	sys::state& state,
+	dcon::nation_id s,
+	dcon::commodity_id c
+) {
+	auto total = 0.f;
+	state.world.nation_for_each_state_ownership(s, [&](auto soid) {
+		auto sid = state.world.state_ownership_get_state(soid);
+		auto market = state.world.state_instance_get_market_from_local_market(sid);
+
+		auto local = state.world.market_get_stockpile(market, c);
+		total += local;
+	});
+
+	return total;
+}
+float market_pool(
+	sys::state& state,
+	dcon::commodity_id c
+) {
+	auto total = 0.f;
+	state.world.for_each_market([&](auto m) {
+		auto local = state.world.market_get_stockpile(m, c);
+		total += local;
+	});
+	return total;
+}
+
+float demand_satisfaction(
+	sys::state& state,
+	dcon::market_id s,
+	dcon::commodity_id c
+) {
+	auto d = demand(state, s, c);
+	auto con = consumption(state, s, c);
+	if(d == 0.f) {
+		return 0.f;
+	}
+	return con / d;
+}
+float demand_satisfaction(
+	sys::state& state,
+	dcon::nation_id s,
+	dcon::commodity_id c
+) {
+	auto d = demand(state, s, c);
+	auto con = consumption(state, s, c);
+	if(d == 0.f) {
+		return 0.f;
+	}
+	return con / d;
+}
+float demand_satisfaction(
+	sys::state& state,
+	dcon::commodity_id c
+) {
+	auto d = demand(state, c);
+	auto con = consumption(state, c);
+	if(d == 0.f) {
+		return 0.f;
+	}
+	return con / d;
 }
 
-void register_intermediate_demand(sys::state& state, dcon::nation_id n, dcon::commodity_id commodity_type, float amount, economy_reason reason) {
-	register_demand(state, n, commodity_type, amount, reason);
-	state.world.nation_get_intermediate_demand(n, commodity_type) += amount;
 
-	float price = state.world.commodity_get_current_price(commodity_type);
-	float sat = state.world.nation_get_demand_satisfaction(n, commodity_type);
+float average_capitalists_luxury_cost(
+	sys::state& state,
+	dcon::nation_id s
+) {
+	auto total = 0.f;
+	auto count = 0.f;
+	state.world.nation_for_each_state_ownership(s, [&](auto soid) {
+		auto sid = state.world.state_ownership_get_state(soid);
+		auto market = state.world.state_instance_get_market_from_local_market(sid);
+
+		auto local_count = state.world.state_instance_get_demographics(
+			sid,
+			demographics::to_key(state, state.culture_definitions.capitalists)
```

Comment:
Since the key is a constant value here, you could move the function that computes it out of the loop.